pyrefly/lib/state/lsp.rs

```diff
@@ -2270,6 +2276,193 @@ impl<'a> Transaction<'a> {
+        position: TextSize,
+        completions: &mut Vec<CompletionItem>,
+    ) {
+        let Some((subscript, string_lit)) = Self::subscript_string_literal_at(module, position)
+        else {
+            return;
+        };
+        let literal_range = string_lit.range();
+        // Allow the cursor to sit a few characters before the literal (e.g. between nested
+        // subscripts) so completion requests fired just before the quotes still succeed.
+        let allowance = TextSize::from(4);
+        let lower_bound = literal_range
+            .start()
+            .checked_sub(allowance)
+            .unwrap_or_else(|| TextSize::new(0));
+        if position < lower_bound || position > literal_range.end() {
+            return;
+        }
+        let base_expr = subscript.value.as_ref();
+        let mut suggestions: BTreeMap<String, Option<Type>> = BTreeMap::new();
+
+        if let Some(bindings) = self.get_bindings(handle) {
+            let base_info = if let Some((identifier, facets)) = Self::expression_facets(base_expr) {
+                Some((identifier, facets))
+            } else if let Expr::Name(name) = base_expr {
+                Some((Ast::expr_name_identifier(name.clone()), Vec::new()))
+            } else {
+                None
+            };
+
+            if let Some((identifier, facets)) = base_info {
+                let short_id = ShortIdentifier::new(&identifier);
+                let idx_opt = {
+                    let bound_key = Key::BoundName(short_id);
+                    if bindings.is_valid_key(&bound_key) {
+                        Some(bindings.key_to_idx(&bound_key))
+                    } else {
+                        let def_key = Key::Definition(short_id);
+                        if bindings.is_valid_key(&def_key) {
+                            Some(bindings.key_to_idx(&def_key))
+                        } else {
+                            None
+                        }
+                    }
+                };
+
+                if let Some(idx) = idx_opt {
+                    let facets_clone = facets.clone();
+                    if let Some(keys) = self.ad_hoc_solve(handle, |solver| {
+                        let info = solver.get_idx(idx);
+                        info.key_facets_at(&facets_clone)
+                    }) {
+                        for (key, ty_opt) in keys {
+                            suggestions.entry(key).or_insert(ty_opt);
+                        }
+                    }
+                }
+            }
+        }
+
+        if let Some(base_type) = self.get_type_trace(handle, base_expr.range()) {
+            if let Some(typed_keys) = self.collect_typed_dict_keys(handle, base_type) {
+                for (key, ty) in typed_keys {
+                    let entry = suggestions.entry(key).or_insert(None);
+                    if entry.is_none() {
+                        *entry = Some(ty);
+                    }
+                }
+            }
+        }
```

Comment:
nit: combine branches

```suggestion
        if let Some(base_type) = self.get_type_trace(handle, base_expr.range())
          && let Some(typed_keys) = self.collect_typed_dict_keys(handle, base_type) {
              for (key, ty) in typed_keys {
                  let entry = suggestions.entry(key).or_insert(None);
                  if entry.is_none() {
                      *entry = Some(ty);
                  }
              }
        }
```